x/nexus/keeper/general_message.go

```diff
@@ -272,3 +268,90 @@ func (k Keeper) GetProcessingMessages(ctx sdk.Context, chain exported.ChainName,
 		return funcs.MustOk(k.GetMessage(ctx, id))
 	})
 }
+
+// SetNewMessage_ sets the given general messsage as approved
+func (k Keeper) SetNewMessage_(ctx sdk.Context, msg exported.GeneralMessage) error {
+	if _, ok := k.GetMessage(ctx, msg.ID); ok {
+		return fmt.Errorf("general message %s already exists", msg.ID)
+	}
+
+	if !msg.Is(exported.Approved) {
+		return fmt.Errorf("new general message has to be approved")
+	}
+
+	funcs.MustNoErr(ctx.EventManager().EmitTypedEvent(&types.MessageReceived{
+		ID:          msg.ID,
+		PayloadHash: msg.PayloadHash,
+		Sender:      msg.Sender,
+		Recipient:   msg.Recipient,
+	}))
+
+	return k.setMessage(ctx, msg)
+}
+
+// SetMessageProcessing_ sets the given general message as processing and perform
+// validations on the message
+func (k Keeper) SetMessageProcessing_(ctx sdk.Context, id string) error {
+	msg, ok := k.GetMessage(ctx, id)
+	if !ok {
+		return fmt.Errorf("general message %s not found", id)
+	}
+
+	if !msg.Is(exported.Approved) && !msg.Is(exported.Failed) {
+		return fmt.Errorf("general message has to be approved or failed")
+	}
+
+	if err := k.validateMessage(ctx, msg); err != nil {
+		return err
+	}
+
+	msg.Status = exported.Processing
+	if err := k.setMessage(ctx, msg); err != nil {
+		return err
+	}
+
+	funcs.MustNoErr(k.setProcessingMessageID(ctx, msg))
+	funcs.MustNoErr(ctx.EventManager().EmitTypedEvent(&types.MessageProcessing{ID: msg.ID}))
+
+	return nil
+}
+
+func (k Keeper) validateMessage(ctx sdk.Context, msg exported.GeneralMessage) error {
```

Comment:
this function is the linchpin of the whole change. Can you add explanations for the checks?